User provided patterns for inline function mocking

laurensmiers opened the pull request:
Proposal for #264 
Yet another pull request for the inline function mocking (I will stop at one point, I promise :)...)
A new configuration option `:inline_function_patterns` is added to make the inline function  recognition fully user-configurable. We provide defaults which are the old ones I had previously hardcoded in the header_parser.

The user does have to keep in mind the patterns he provides are directly fed to a Regexp and thus must be properly escaped etc.
I tried adding `Regexp.escape` to remove problem characters, but this conflicts with actual regex string patterns we don't want escaped (f.e. `\b` disappears) and anyway, I think providing actual regex patterns is in line with the way f.e. `:strippables` works so I guess users are used to this by now. (correct me if I'm wrong)

KR,

Laurens

Letme commented:
I think the error I am seeing with testing your branch (btw thanks, the regexp works) is that you include the original header (which has static inline implementation) in the mock header. Then I get redefinition error (keep in mind we have our own build system, so that might contribute to problems we are having). Sounds possible that you have something else?

Letme commented:
Here is a short example, I don't know how you solved, but we have it:

example.h:
```
static inline int myfunc(void) {
    return 0;
}
```
generates mock_example.h:
```
#include "example.h" /* <--- this presents a problem now, as this includes
                      * inline definition of the static inline function */

mock_example_Init();
mock_example_Verify();
....
```


generates mock_example.c:
```
#include "mock_example.h"

int myfunc(void)
{
    ...
    CMOCK STUFF
    ...
}
```

Letme commented:
OK, here is how I am solving it now:

Here is a short example, I don't know how you solved, but we have it:

example.h:
```
#ifndef UNITTEST
static inline int myfunc(void) {
    return 0;
}
#endif
```
generates mock_example.h:
```
#include "example.h" /

mock_example_Init();
mock_example_Verify();
....
```


generates mock_example.c:
```
#include "mock_example.h"

int myfunc(void)
{
    ...
    CMOCK STUFF
    ...
}
```

laurensmiers commented:
@Letme Thanks for testing and the feedback. There is some information missing from the documentation that should help with the issue you are seeing, sorry I forgot to add it...
I'll add it in this pull request.
We are generating a new header "example.h" next to mock_example.h/c with the same name as the original to-be-mocked header, but with the inline functions transformed to normal functions.
This makes sure the mock_example.h/c include our new generated header example.h.
So we are still adding `#include "example.h"` in the mock_example.h but this should no longer be the original "example.h", instead it should be the generated one that is included.

But this does mean that for the compilation of the tests, you probably have to add the folder where your mocks (something like build/test/mocks for Ceedling)are being stored to your include paths. Also mind the order of include paths, you should set the mock/build folder before the other includes to avoid that you include the original header, which is what you are seeing I think.

Letme commented:
Ok. that makes sense = modified our build system accordingly.

I have one more strange occurence I cant explain.

original example.h:
```
static __inline__ __attribute__ ((always_inline)) uint16_t _somefunc (uint32_t a)
{
    return _someotherfunc (a);
}

static __inline__ uint16_t _somefunc_0  (uint32_t a)
{
    return (uint16_t) a;
}
```

cmock generated example.h:
```
uint16_t _somefunc (uint32_t a);

static __inline__ uint16_t _somefunc_0  (uint32_t a);
```

CMockConfig.yml:
```
...
    :inline_function_patterns:
        - 'static inline'
        - 'static __inline__ __attribute__ \(\(always_inline\)\)'
        - 'static __inline__'
        - 'static __attribute__\(\(always_inline\)\) inline'
...
```

laurensmiers commented:
I took your example.h and the :inline_function_patterns but I can't seem to reproduce it, not in the unit tests or with my example project. 

If I take your example.h in my example project, mock_example.h and the new example.h contain the correct function declarations. I do however have problems with cmock_example.c where the return value is `__inline__ uint16_t` instead of `uint16_t` for `_somefunc_0`. I can fix that by taking the `:inline_function_patterns` into account when making the mocks so that's something we can fix but it's a different issue.

I hate to ask because I know it can be a pain to extract (and sometimes the issue dissappears if you extract part of the code...) or impossible due to legal stuff but would it be possible to provide an example project that has the issue you are seeing? If you don't have time, no problem.
Either way, I'll make a patch with some debug info to locate where we miss the `_somefunc_0`

laurensmiers commented:
I pushed a new branch with some debug prints added: https://github.com/laurensmiers/CMock/tree/fix_bug_pull_request 
If you have time, could you run your example.h with this branch and post the output when running cmock?
If we can find it like this, you can skip making the example project and we save time :).

Letme commented:
That output was indeed helpful as I cant share the code, but it got me to conclusion that we have  a following case in example.h:
```
static __inline__ uint16_t _somefunc (uint32_t a) __attribute__ ((always_inline));

static __inline__ uint16_t _somefunc (uint32_t a)
{
    return _someotherfunc (a);
}
```
That lead to correct mock of sumefunc, but the first definition was just copied over (so attribute on end remained).

What is even worse in same library we have:
```
__attribute__((deprecated, always_inline))
static inline void somefunc (void);

static inline void somefunc (uint32_t a)
{
    return 5;
}
```

Anyway a bit out of scope with both, so I am wrapping definitions with `ifndef TEST` to get over it as especially 2nd case is something of "exception".

Letme commented:
I turned back `:weak` on (since we want mocks to be weak) and I am back at issue where two styles in same file produce that in second part the `static __inline__` is not stripped. I have my questions about the order in which we have them defined, so I will play a bit with that

This example.h

```
static __inline__ __attribute__ ((always_inline)) uint16_t _somefunc (uint32_t a);

static __inline__ __attribute__ ((always_inline)) uint16_t _somefunc (uint32_t a)
{
    return _someotherfunc (a);
}

static __inline__ uint16_t _somefunc_0  (uint32_t a)
{
    return (uint16_t) a;
}
```

Letme commented:
So above example might hold the key why its not wroking and why it is. Here is a more accurate example of the latest example.h:

```
#ifndef TEST
static __inline__ uint16_t _somefunc_0 (uint32_t a) __attribute__ ((always_inline));
#endif

static __inline__ __attribute__ ((always_inline)) uint16_t _somefunc (uint32_t a)
{
    return _somefunc_0 (a);
}

static __inline__ uint16_t _somefunc_0  (uint32_t a)
{
    return (uint16_t) a;
}
```

is generated into:
```
uint16_t _somefunc (uint32_t a);

static __inline__ uint16_t _somefunc_0  (uint32_t a);
```

and this causes an error when we add `:weak: - '__attribute__ ((weak))'`

laurensmiers commented:
> That output was indeed helpful as I cant share the code, but it got me to conclusion that we have a following case in example.h:
> 
> ```
> static __inline__ uint16_t _somefunc (uint32_t a) __attribute__ ((always_inline));
> 
> static __inline__ uint16_t _somefunc (uint32_t a)
> {
>     return _someotherfunc (a);
> }
> ```
> 
> That lead to correct mock of sumefunc, but the first definition was just copied over (so attribute on end remained).
> 
> What is even worse in same library we have:
> 
> ```
> __attribute__((deprecated, always_inline))
> static inline void somefunc (void);
> 
> static inline void somefunc (uint32_t a)
> {
>     return 5;
> }
> ```
> 
> Anyway a bit out of scope with both, so I am wrapping definitions with `ifndef TEST` to get over it as especially 2nd case is something of "exception".

Wow, that is .... interesting (let's call it that ;)). Yeah, not sure how we could handle this in a generic way. Maybe it's worth making a separate issue for this, just so we don't forget and maybe we can find some generic way in the future... cause right now I fail to see a way except your workaround of surrounding it with `#ifndef TEST ... #endif`

laurensmiers commented:
> So above example might hold the key why its not wroking and why it is. Here is a more accurate example of the latest example.h:
> 
> ```
> #ifndef TEST
> static __inline__ uint16_t _somefunc_0 (uint32_t a) __attribute__ ((always_inline));
> #endif
> 
> static __inline__ __attribute__ ((always_inline)) uint16_t _somefunc (uint32_t a)
> {
>     return _somefunc_0 (a);
> }
> 
> static __inline__ uint16_t _somefunc_0  (uint32_t a)
> {
>     return (uint16_t) a;
> }
> ```
> 
> is generated into:
> 
> ```
> uint16_t _somefunc (uint32_t a);
> 
> static __inline__ uint16_t _somefunc_0  (uint32_t a);
> ```
> 
> and this causes an error when we add `:weak: - '__attribute__ ((weak))'`

Just for my understanding so I can try to reproduce:
Your current cmock.yml is:
```
    :inline_function_patterns:
        - 'static inline'
        - 'static __inline__ __attribute__ \(\(always_inline\)\)'
        - 'static __inline__'
        - 'static __attribute__\(\(always_inline\)\) inline'
```

The header is the one you listed above ,with the `TEST` macro defined to get rid of the first declaration.

If you disable `:weak` the example.h header is generated correctly like this:

```
uint16_t _somefunc (uint32_t a);

uint16_t _somefunc_0  (uint32_t a);
```

But if you enable `:weak`, the `static __inline__` is **not** stripped from _somefunc_0.

Is this correct or have I misunderstood?

Letme commented:
I think you had it correct originally, but i do not know if weak highlights the issue or just produces it. I will do some more checking tomorrow (although you can remind me next week as i should have more time then)

laurensmiers commented:
Ok, thanks! let me know if you find something. Meanwhile, I'll try to reproduce on my side with some more tests.
I would also check the debug output to see if the `static __inline__` is picked up by CMock. (see my EDIT 2 in the message above what the expected output should be)
Because from the looks of it, it isn't but I can't figure out why... And I suspect that even with `:weak` disabled, the `static __inline__` is not stripped from the generated header. The `:weak` config option should have no impact on the generated example.h header if I look at the code (I could be missing something, will check further if there is something linking them).
Last thing I can think of is playing with the order of the `:inline_function_patterns`, if you remove all patterns and only specify `static __inline__`, it should only touch `_somefunc_0` and make that a 'normal' function, maybe that is something interesting to test? That would indicate a (hidden) dependency in the `:inline_function_patterns`, it's just a guess from my side.

Letme commented:
I have indeed played with the order earlier and if I put `static __inline__` before the one with attribute then it does strip the whole file of `static __inline__`, but compilation fails because top functions remain having the attribute...

Can you refresh the debug printout branch with latest changes from this branch as well?

@markvrvoord I do not think that this is something blocking this MR, so for me it can go in asap, because it does have added value.

Letme commented on test/unit/cmock_header_parser_test.rb:
```diff
@@ -1997,4 +2029,31 @@
+      "	b -= 32;\n" +
+      "	return a + (int)(b) + (int)c;\n" +
+      "}\n" +
+      "static __inline__ __attribute__ ((always_inline)) uint16_t attributealwaysinlinefuncname(void) {\n" + # Second user pattern
```
Maybe order here is important for this to fail. Can you put this before the `static __inline__` only, so that we see if it fails in that combination?

laurensmiers replied:
I tried to move the `attributealwasyinlinefuncname` before `dummy_func_2` but the test still passes.
As stated in the MR, if I move `static __inline__` up in the array declared on line 2055:
```
@parser.inline_function_patterns = ['static __inline__', 'static __inline__ __attribute__ \(\(always_inline\)\)', 'static inline']
```
Then the test fails because `__attribute__((always_inline))` is not removed but it's still another error than the one you are seeing I think.

Letme replied:
Ok, if test still passes then there is something more with my issue. Anyway - it is not blocking merge of this.

laurensmiers commented:
> I have indeed played with the order earlier and if I put `static __inline__` before the one with attribute then it does strip the whole file of `static __inline__`, but compilation fails because top functions remain having the attribute...
> 
> Can you refresh the debug printout branch with latest changes from this branch as well?
> 
> @mvandervoord I do not think that this is something blocking this MR, so for me it can go in asap, because it does have added value.

I updated the debug printout branch, hope it gives some more info...

And yes, If we have 
```
    :inline_function_patterns:
        - 'static inline'
        - 'static __inline__' <----------- Moved up
        - 'static __inline__ __attribute__ \(\(always_inline\)\)'
        - 'static __attribute__\(\(always_inline\)\) inline'
```
Then we will miss the functions with the `__attribute__((always_inline))` since we already removed `static __inline__` and we are checking for `static __inline__ __attribute__((always_inline))` only after.
The order for `:inline_function_patterns` should go from specific to general for this reason (updated documentation to state this).

But I don't think this is the cause of the issue you are seeing since the :inline_function_patterns you're using looks ok (last version that I know of):
```
    :inline_function_patterns:
        - 'static inline'
        - 'static __inline__ __attribute__ \(\(always_inline\)\)'
        - 'static __inline__'
        - 'static __attribute__\(\(always_inline\)\) inline'
```
And the `static __inline__` is getting removed if you move it up so it is getting picked up I guess.
So I'm starting to think it may be something that you stated earlier, is there somewhere still a hidden forward declaration of the `static __inline__ _somefunc_0` that gets copied over somehow? Though I would expect it to give other compile errors/warnings if that was the case but who knows :)

Letme commented:
When I open the generated file it has functions with stripped `static __inline__ __attribute__ ((always_inline))` and functions which still contain the `static __inline__` . My idea was that after first hit already removes the functions and that afterwards the subsequent removals are not triggered, but your test case proves that this is not the case. Might there be count issue? 

We have around 17 functions with `static __inline__ __attribute__ ((always_inline))` before someone decided that `static __inline__` is sufficient (and we then have around 16 functions with `static __inline__` - all of them are unstripped).

laurensmiers commented:
Hmmm it does sound like it stops after the first regex is completely done, but I can't reproduce this in the unit tests. 
My proposal if you're ok with that: I'll make a dummy project/repo with my debug branch and Ceedling (just because it's quick to setup) with a header and CMock configuration as close to your project as possible and best case, we can trigger the issue, or worst case, we have a working header generator which maybe can pinpoint what is causing the issue in your project.
Because I'm not fully comfortable with the inline mocking working in our tests but not entirely working in a real project. But we can of course try to merge this and look into the issue in parallel, up to @mvandervoord to decide :).

mvandervoord commented:
If you've made progress already and haven't broken anything non-inline related, I'm happy to merge.

Is debug code all back out of this branch?

laurensmiers commented:
All debug code is out of this branch + removed the double unit test that snuck in when debugging.
If something comes out of the testing with @Letme's project, that will be a separate MR I guess.
This one seems to be ok for the current known (inline) usecases and doesn't seem to have broken antyhing non-inline related (I'll hear about it when it does I guess :))
